src/lib/repositories/impl_oauth2/oauth2_repository_impl.py

```diff
@@ -0,0 +1,257 @@
+import bcrypt
+from injector import inject
+from sqlalchemy import text
+from sqlalchemy.engine.base import Engine
+import jwt
+
+from src.constants.http_status_code import HttpStatus
+from src.constants.oauth2 import CredentialType
+from src.env_config import get_env_config_instance
+from src.utils.oauth2_util import create_token
+from src.utils.sql_oath_queries import (
+    SQL_QUERY_TO_GET_USER_BY_USER_NAME,
+    SQL_QUERY_TO_GET_CLIENT_BY_CLIENT_ID,
+    SQL_QUERY_TO_GET_CLIENT_BY_ID,
+    SQL_QUERY_TO_ADD_ACCESS_TOKEN,
+    SQL_QUERY_TO_ADD_REFRESH_TOKEN,
+    SQL_QUERY_TO_GET_REFRESH_TOKEN_BY_TOKEN,
+    SQL_QUERY_TO_GET_ACCESS_TOKEN_BY_TOKEN,
+    SQL_QUERY_TO_GET_REFRESH_TOKEN_BY_ID,
+    SQL_QUERY_TO_DELETE_ACCESS_TOKEN,
+    SQL_QUERY_TO_DELETE_REFRESH_TOKEN,
+    SQL_QUERY_TO_GET_SCOPE_BY_CLIENT_ID,
+)
+
+
+class Oauth2Repository:
+    @inject
+    def __init__(self, engine: Engine):
+        self.env_config = get_env_config_instance()
+        self.engine = engine
+
+    def login(
+        self, client_id, client_secret, credential_type, user_name=None, password=None
+    ):
+        client = self._get_client_by_client_id(client_id)
+
+        if client.client_secret == client_secret:
+
+            if credential_type == CredentialType.CLIENT_CREDENTIAL:
+                tokens = self._client_credential_login(client)
+
+                return tokens
+
+            else:
+                tokens = self._client_user_credential_login(client, user_name, password)
+                return tokens
+        return HttpStatus.UNAUTHORIZED.value
+
+    def validate_token(self, token):
+        try:
+            decoded_token = jwt.decode(
+                token, self.env_config.secret_key, algorithms="HS256"
+            )
+
+            return {"access_token": token}
+
+        except jwt.ExpiredSignatureError:
+
+            access_token = self._get_access_token_from_db(token)
+
+            refresh_token = self._get_refresh_token_by_id(access_token.refresh_token_id)
+            self._delete_access_token(access_token.token)
+            try:
+                decoded_refresh_token = jwt.decode(
+                    refresh_token.token,
+                    self.env_config.secret_key,
+                    algorithms="HS256",
+                )
+                new_access_token = self._refresh_token(
+                    decoded_refresh_token, refresh_token.id
+                )
+                return new_access_token
+
+            except jwt.ExpiredSignatureError:
+                self._delete_refresh_token(refresh_token.token)
+                return None
+
+        except jwt.InvalidSignatureError:
+            return None
+        except jwt.DecodeError:
+            return None
+
+    def _refresh_token(self, decoded_refresh_token, refresh_token_id):
+        client = self._get_client_by_id(decoded_refresh_token.get("client_id"))
+
+        if decoded_refresh_token.get("user_name"):
+
+            new_access_token = create_token(
```

Comment:
Create a function to build the new access token

Reply:
done.